<!-- Dependencies
1. Libraries:
- biomaRt
- dplyr
- goseq (getlengths)

2. Libraries (human-specific)
- Homo.sapiens
- TxDb.Hsapiens.UCSC.hg38.knownGene
-->

### Load gene annotations

Next, we will use [Bioconductor
OrganismDb](http://bioconductor.org/packages/release/bioc/html/OrganismDbi.html)
packages `citep(citation('OrganismDbi'))` and the R interface to BioMart 
`citep(c('10.1186/1471-2164-10-22', '10.1038/nprot.2009.97'))` to retrieve
ENSEMBL gene annotations for the host.

```{r load_gene_annotations, message=FALSE, warning=FALSE, error=FALSE, results='asis'}
# Load gene and transcript annotations
library(CONFIG$organism_db, character.only=TRUE)
orgdb <- get(CONFIG$organism_db)

# Fix AnnotationDbi namespace mess
assign('select', dplyr::select, envir=.GlobalEnv)
assign('get',    base::get, envir=.GlobalEnv)

# main ensembl host is down fairly often, so alt hosts may be needed.
gene_info <- load_host_annotations(orgdb, rownames(count_table),
                                   keytype=CONFIG$orgdb_key,
                                   biomart_dataset=CONFIG$biomart_dataset,
                                   biomart_host='uswest.ensembl.org')
                                   #biomart_host='www.ensembl.org')
                                   #biomart_host='aug2017.archive.ensembl.org')
                                   #biomart_host='archive.ensembl.org')

# Keep only the feature information remaining genes
gene_info <- gene_info[gene_info$gene_id %in% rownames(count_table),]

# For now, just grab the description for the first transcript
gene_info <- gene_info[!duplicated(gene_info$gene_id), ]

# Gene IDs
gene_ids <- rownames(count_table)

# Determine approximate transcript lengths for each gene.
# Note that because the annotations have not been filtered to remove ncRNAs,
# many of the genes associated with lncRNAs, etc. will not have a length
# associated with them.
gene_info$transcript_length <- getlength(gene_ids, CONFIG$organism_genome, "ensGene")

# Note 2017/12/09
# ---------------
# hg19 still better to use for tx lengths than hg38 for the moment:
# > sum(is.na(getlength(gene_ids, 'hg38', "ensGene"))                                                                                                                                                            
#   [1] 36591
# > sum(is.na(getlength(gene_ids, 'hg19', "ensGene")))                                                                                                                                                           
#   [1] 24462      

# Location of external annotation files
species_dir <- tolower(sub('. ', '', CONFIG$host))
input_dir <- file.path('..', '..', 'data', species_dir)
```

```{r results='asis', include=CONFIG$verbose, echo=CONFIG$verbose}
# gene annotations
if (CONFIG$include_tables) {
  kable(head(gene_info), caption='Preview of gene annotations.')
}
```

### Load GO annotations

Next, we will use the [GO.db annotation
package](http://bioconductor.org/packages/release/data/annotation/html/GO.db.html) 
`citep(citation('GO.db'))` to load a table of Gene Ontology (GO) 
`citep('10.1038/75556')` terms associated with each parasite gene.

```{r load_go_annotations, message=FALSE, warning=FALSE, error=FALSE, results='asis'}
# Load GO annotations (for host species, exclude ancestor terms to speed things
# up; goseq will handle this for us later on)
go_terms <- load_go_terms(orgdb, rownames(count_table), keytype=CONFIG$orgdb_key,
                          include_ancestors=FALSE)

# Gene / GO term mapping
gene_go_mapping <- as.data.frame(unique(
  go_terms %>% select(.data[[CONFIG$orgdb_key]], GO, ONTOLOGY)
))
colnames(gene_go_mapping) <- c('gene', 'category', 'ontology')
go_term_id_mapping <- as.data.frame(unique(go_terms[c('GO', 'TERM', 'ONTOLOGY')]))
colnames(go_term_id_mapping) <- c("category", "term", "ontology")
```

```{r results='asis', include=CONFIG$verbose, echo=CONFIG$verbose}
# Preview of GO term annotations
if (CONFIG$include_tables) {
  kable(head(go_terms), caption='Preview of GO annotations.')
}
```

### Load KEGG annotations

Next, we will load a table of KEGG `citep('10.1093/nar/27.1.29')` pathway
annotations for each host gene.

```{r load_kegg_annotations}
# For mouse/human, KEGG mapping has to be loaded separately

# Human
if (CONFIG$host == 'H. sapiens') {
  kegg_mapping_file  <- file.path(input_dir, 'Hsapiens_KEGG_Annotations.csv')
  kegg_pathways_file <- file.path(input_dir, 'Hsapiens_KEGG_Pathways.csv')
  org_abbreviation <- 'hsa'
} else if (CONFIG$host == 'M. musculus') {
  # Mouse
  kegg_mapping_file  <- file.path(input_dir, 'Mmusculus_KEGG_Annotations.csv')
  kegg_pathways_file <- file.path(input_dir, 'Mmusculus_KEGG_Pathways.csv')
  org_abbreviation <- 'mmu'
}

if (file.exists(kegg_mapping_file)) {
  # If KEGG mapping are available, load from file
  gene_kegg_mapping <- read.csv(kegg_mapping_file)
  kegg_pathways <- read.delim(kegg_pathways_file)
} else {
  # Otherwise use KEGGREST to construct mappings
  library('KEGGREST')

  pathways <- unique(keggLink("pathway", org_abbreviation))
  kegg_pathways <- generate_kegg_pathway_mapping(pathways, CONFIG$verbose)
  gene_kegg_mapping <- generate_gene_kegg_mapping(pathways,
                                                  org_abbreviation,
                                                  verbose=CONFIG$verbose)

  # Save KEGG mapping
  if(!file.exists(input_dir)) {
      dir.create(input_dir, recursive=TRUE)
  }

  write.csv(gene_kegg_mapping, file=kegg_mapping_file, quote=FALSE,
          row.names=FALSE)
  write.table(kegg_pathways, file=kegg_pathways_file, quote=FALSE,
          row.names=FALSE, sep='\t')
}

# Rename gene/KEGG mapping columns to be consistent with GO mapping
colnames(gene_kegg_mapping) <- c('gene', 'category')
colnames(kegg_pathways)     <- c('category', 'name', 'class', 'description')

kegg_pathways <- unique(kegg_pathways)
```
